backpropagation/network/neural_network.py

```diff
@@ -122,43 +129,45 @@ def _backpropagation(
         activation_function_arguments, activation_function_values = (
             self._collect_activation_function_arguments_and_values(x))
 
-        output_error = np.multiply(
-            self.cost_function.calculate_derivative_value(
-                y, activation_function_values[-1]),
-            self.activation_function.calculate_derivative_value(
-                activation_function_arguments[-1]),
-        )
-
-        gradient = list()
-        error = output_error
-        weight_gradient = (
-            np.atleast_2d(activation_function_values[-2]).transpose() @
-            np.atleast_2d(error)
-        )
-
-        gradient.append(
-            np.insert(weight_gradient.transpose(), 0, error, axis=1))
-        for layer in range(2, self.layers_count):
+        def propagate_backwards(
+                errors: Tuple[np.array, np.array],
+                layer: int
+        ) -> Tuple[np.array, np.array]:
+            error, _ = errors
             activation_derivative = self.activation_function.calculate_derivative_value(
-                activation_function_arguments[-layer]
-            )
+                activation_function_arguments[-layer])
 
-            weight = self.weights[-layer + 1][:, 1:]
-            error = weight.transpose() @ error * activation_derivative
+            if layer == 1:
+                cost_derivative = self.cost_function.calculate_derivative_value(
+                    y, activation_function_values[-layer])
+                error = activation_derivative * cost_derivative
+            else:
+                weight = self.weights[-layer + 1][:, 1:]
+                error = weight.transpose() @ error * activation_derivative
 
             weight_gradient = (
-                np.atleast_2d(activation_function_values[-layer - 1]).transpose() @
-                np.atleast_2d(error)
+                error @
+                activation_function_values[-layer - 1].T
             )
-            gradient.insert(0, np.insert(
-                weight_gradient.transpose(), 0, error, axis=1))
 
+            bias_gradient = np.sum(error, axis=1, keepdims=True)
+            gradient = np.hstack((bias_gradient, weight_gradient)) / x.shape[1]
+
+            return error, gradient
+
+        deltas = list(accumulate(
+            [(None, None)] + list(range(1, self.layers_count)),
+            propagate_backwards
+        ))
+
+        gradient = [d[1] for d in deltas[1:]]
+        gradient.reverse()
```

Comment:
Seems costly

Reply:
Fixed in ec76e254beec8723a922814b2492ace1fa938ed9

Switched to iterating in reverse.